crates/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -778,12 +778,10 @@ impl Binary {
                 let rhs = truncate(rhs.try_into_u128()?, *bit_size);
 
                 // The divisor is being truncated into the type of the operand, which can potentially
-                // lead to the rhs being zero.
-                // If the rhs of a division is zero, attempting to evaluate the divison will cause a compiler panic.
-                // Thus, we do not evaluate the division in this method, as we want to avoid triggering a panic,
-                // and the operation should be handled by ACIR generation.
+                // lead to the rhs being zero. This should be caught before instruction simplification
+                // and passing this if statement is a compiler bug.
                 if matches!(self.operator, BinaryOp::Div) && rhs == 0 {
-                    return None;
+                    unreachable!("ICE: the divisor of a binary op has been truncated into zero");
```

Comment:
I was thinking this was going to trigger if we did `let z: u32 = 3 / 0`, but it does not for some reason. Do you know why?

Reply:
`eval_constants` is only [called](https://github.com/noir-lang/noir/blob/0920dd03d67c50da36bfb87db2e50f6a4aa155bd/crates/noirc_evaluator/src/ssa/ir/instruction.rs#L638)  after a check whether the rhs is `0` and we are not attempting a div. The truncate inside of `eval_constants` was truncating a value into the type of the lhs. In the case of `let a: u1 = 1 >> 1` we are attempting to do `1 / 2`, and thus are attempting to truncate 2 into a `u1` which is why the rhs would be 0. For `let z: u32 = 3 / 0` we never hit `eval_constants`.

Reply:
I understand that - I was missing that we actually have another of the same check on line 635 that checks if we're dividing with an rhs of zero. That prevents eval_constants from being called. Perhaps we only need one of these two checks.

Reply:
Yeah we should really only need the first one on 635. It was a mistake to originally have this check inside `eval_constants`, but I thought to leave it as a compiler sanity check for a more clear error than the panic shown in #2250. If we deem the check unnecessary I could just leave a comment stating that if we hit a divide by zero panic it is most likely due to a missing type check for an overflow.